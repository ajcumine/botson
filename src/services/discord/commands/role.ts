import { Message } from 'discord.js';
import { Command } from '.';

export const execute = (message: Message, args: string[]) => {
  const taggedUser = message.mentions.users.first();
  const selectedRole = args.slice(1).join(' ');
  const guildRoles = message.guild.roles;
  if (guildRoles.find(gr => gr.name === selectedRole)) {
    message.channel.send(`You tried to set ${taggedUser} to ${selectedRole}`);
  } else {
    message.reply(
      `the role ${selectedRole} does not exist, available server roles are: ${guildRoles
        .map(gr => `${gr.name}`)
        .join(', ')}`,
    );
  }
};

const role: Command = {
  execute,
  argsRequired: true,
  description: 'Change a users role',
  usage: '<user> <role>',
  guildOnly: true,
};

export default role;
