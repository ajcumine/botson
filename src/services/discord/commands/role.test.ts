import Discord from 'discord.js';
import role, { execute } from './role';

describe('!role command', () => {
  it('handles !role commands', () => {
    expect(role).toMatchSnapshot();
  });

  describe('execute', () => {
    describe('if the role exists', () => {
      it('responds with a role switch message', () => {
        const messageSendSpy = jest.fn();
        const mockMessage = {
          channel: {
            send: messageSendSpy,
          },
          guild: {
            roles: [{ name: 'knownRole' }],
          },
          mentions: {
            users: {
              first: () => 'User',
            },
          },
        };

        execute(mockMessage, ['@User', 'knownRole']);
        expect(messageSendSpy).toHaveBeenCalledWith(
          'You tried to set User to knownRole',
        );
      });
    });

    describe('if the role does not exist', () => {
      it('responds with the available roles', () => {
        const replySpy = jest.fn();
        const mockMessage = {
          guild: {
            roles: [{ name: 'knownRole' }],
          },
          mentions: {
            users: {
              first: () => 'User',
            },
          },
          reply: replySpy,
        };

        execute(mockMessage, ['@User', 'unknownRole']);
        expect(replySpy).toHaveBeenCalledWith(
          'the role unknownRole does not exist, available server roles are: knownRole',
        );
      });
    });
  });
});
