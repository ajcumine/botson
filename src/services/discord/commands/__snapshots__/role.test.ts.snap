// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`!role command handles !role commands 1`] = `
Object {
  "argsRequired": true,
  "description": "Change a users role",
  "execute": [Function],
  "guildOnly": true,
  "usage": "<user> <role>",
}
`;
